### PR TITLE
docs: add conc playback docs to getting started tutorial

### LINKED ISSUE #1439: Write concrete playback tutorial page

### PR DESCRIPTION
**Pull request by sanjit-bhat**
### Description of changes: 

See the title. This PR only adds doc sections.

Note that there are other sections where we could potentially add info about this feature (e.g., 5 sections name the `--visualize` feature). However, as per @tedinski's suggestion, we should save those changes for later once we figure out more details about how the overall tutorial flow would be impacted.

### Callouts:

Some of the names used in this PR rely on the renames in #1548.

### Resolved issues:

Resolves #1439, adding a doc section. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: docs/src/SUMMARY.md [hunk 1]
```diff
@@ -5,6 +5,7 @@
     - [Building from source](./build-from-source.md)
   - [Using Kani](./usage.md)
   - [Verification results](./verification-results.md)
+  - [Debugging harnesses](./concrete-playback.md)
```
Comment (celinval) [inline docs/src/SUMMARY.md:8]: Just a suggestion: Debugging verification result?

File: docs/src/concrete-playback.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+To do this, run `git clone https://github.com/model-checking/kani.git`.
+Then, the user needs to add the following lines to their `Cargo.toml` file:
+```toml
+[dev-dependencies]
+kani = { path = "{path_to_kani}/library/kani", features = ["exe_trace"] }
+```
+
+## Usage
+
+Run Kani with the `--gen-exe-trace` flag.
```
Comment (celinval) [inline docs/src/concrete-playback.md]: I think we should rename these flags before pushing this doc. You can either make this change in this PR or in a separate PR that shall be merged before this one.

File: docs/src/concrete-playback.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
+# Concretely playing back proof harnesses
```
Comment (celinval) [inline docs/src/concrete-playback.md:1]: I expect the title here to match the section name: "Debugging harness"

File: docs/src/concrete-playback.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+When the result of a certain check comes back as a `FAILURE`,
+this _experimental_ feature allows users to debug their proof harness and understand what
```
Comment (celinval) [inline docs/src/concrete-playback.md:4]: you should probably say that Kani offers two different paths today.
 - Visualize: Kani provide an HTML text based trace that enumerate the execution steps that led to the property failure.
 - Concrete playback: This is a new _experimental_ feature that generates a unit test that when executed will trigger the failure seen during the verification.

File: docs/src/concrete-playback.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+caused the failing check.
+
```
Comment (celinval) [inline docs/src/concrete-playback.md:6]: Now add:
```md
## Concrete playback
### Setup
`

File: docs/src/concrete-playback.md [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+## Setup
+
+The user needs to have a fairly recent version of the Kani source code somewhere on their computer.
```
Comment (celinval) [inline docs/src/concrete-playback.md:9]: Be specific. This will be added on version 0.9.

File: docs/src/concrete-playback.md [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+To do this, run `git clone https://github.com/model-checking/kani.git`.
```
Comment (celinval) [inline docs/src/concrete-playback.md:10]: Maybe say, this feature require users to declare dependency on the kani crate. BTW, users don't need to do manually clone the repository. Cargo does that for them.

https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#specifying-dependencies-from-git-repositories

File: docs/src/concrete-playback.md (hunk 2 above)
Comment (celinval) [inline docs/src/concrete-playback.md]: In order to enable this feature, run Kani with `--enable-unstable --gen-exe-trace`.

File: docs/src/concrete-playback.md [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+Then, the user needs to add the following lines to their `Cargo.toml` file:
+```toml
+[dev-dependencies]
+kani = { path = "{path_to_kani}/library/kani", features = ["exe_trace"] }
+```
+
+## Usage
+
+Run Kani with the `--gen-exe-trace` flag.
+After verifying the proof harness checks, Kani will extract concrete values for the `kani::any()` variables
```
Comment (celinval) [inline docs/src/concrete-playback.md:20]: Maybe describe what the user will see and not what the tool is doing under the hood. Say something like if there is a failure, Kani will print a test case to the console that includes concrete values that will trigger the failure concretely.

I would suggest that you add an example here of a failure and how the test will look like.

File: docs/src/concrete-playback.md [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+and generate a Rust unit test case.
+This unit test initializes the concrete values and calls the proof harness.
+A user can then either 1) copy this unit test into the same module as their proof harness or
+2) run Kani with the `--add-exe-trace-to-src` flag to have Kani automatically do this.
```
Comment (celinval) [inline docs/src/concrete-playback.md]: After you talk about the testcase structure, I would introduce the new flag. Something like:

> Kani also offer an option to add the test directly to your source code. For that, invoke kani using the `--in-place` flag....

**Review by celinval — COMMENTED** [review]
Great job. I added a few comments. Let me know what you think.

File: docs/src/SUMMARY.md (hunk 1 above)
Comment (sanjit-bhat) [reply docs/src/SUMMARY.md:8]: done

File: docs/src/concrete-playback.md (hunk 2 above)
Comment (sanjit-bhat) [reply docs/src/concrete-playback.md]: done

File: docs/src/concrete-playback.md (hunk 3 above)
Comment (sanjit-bhat) [reply docs/src/concrete-playback.md:1]: done

File: docs/src/usage.md [hunk 10]
```diff
@@ -23,6 +23,8 @@ This works like `cargo test` except that it will analyze all proof harnesses ins
 
 Common to both `kani` and `cargo kani` are many command-line flags:
 
+ * `--concrete-playback=JustPrint`: Generates a Rust unit test case that plays back a failing proof harness using a concrete counterexample. The additional `--concrete-playback=InPlace` flag will automatically add this unit test to the source code next to the proof harness.
```
Comment (celinval) [inline docs/src/usage.md:26]: `--concrete-playback=[print | inplace]`: Generates a Rust unit test that represents a concrete counter-example found by Kani. Use `print` argument to print the test in the terminal or `inplace` to automatically add this unit test to the source code next to the target proof harness.

File: docs/src/concrete-playback.md (hunk 7 above)
Comment (sanjit-bhat) [reply docs/src/concrete-playback.md:14]: Done

File: docs/src/concrete-playback.md (hunk 6 above)
Comment (sanjit-bhat) [reply docs/src/concrete-playback.md]: done

File: docs/src/concrete-playback.md (hunk 5 above)
Comment (sanjit-bhat) [reply docs/src/concrete-playback.md:10]: Done

File: docs/src/concrete-playback.md (hunk 2 above)
Comment (sanjit-bhat) [reply docs/src/concrete-playback.md]: done

File: docs/src/concrete-playback.md (hunk 8 above)
Comment (sanjit-bhat) [reply docs/src/concrete-playback.md]: Done

File: docs/src/concrete-playback.md (hunk 9 above)
Comment (sanjit-bhat) [reply docs/src/concrete-playback.md]: done

File: docs/src/usage.md (hunk 10 above)
Comment (sanjit-bhat) [reply docs/src/usage.md]: done

**Comment by sanjit-bhat** [comment]
@celinval I added an example, but I decided to only explain a couple parts of the unit test that I thought would be most relevant to the user. Here, I define relevance as "info that's helpful if the user wants to try a different concrete value without regenerating the unit test".

File: docs/src/debugging-verification-failures.md [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+```
+Here, `133` and `35207` are the concrete values that, when substituted for `a` and `b`,
+cause an assertion failure.
+`vec![135, 137]` is the byte array representation of `35207`.
+
+### Common issues
+
+* `error[E0425]: cannot find function x in this scope`:
+this is usually caused by having `#[cfg(kani)]` somewhere in the control flow path of the user's proof harness.
+To fix this, remove `#[cfg(kani)]` from those paths.
```
Comment (celinval) [inline docs/src/debugging-verification-failures.md:86]: I think it's better to recommend adding `--cfg=kani` to your rustflags.

File: docs/src/debugging-verification-failures.md [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+
+### Request for comments
+
+This feature is experimental and is therefore subject to change.
+If you have ideas for improving the user experience of this feature,
+please add them to [this github issue](https://github.com/model-checking/kani/issues/1536).
+
+### Limitations
```
Comment (celinval) [inline docs/src/debugging-verification-failures.md:94]: Can you link the existing issues here?

File: docs/src/debugging-verification-failures.md [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+This section describes the concrete playback feature in more detail.
+
+### Setup
+
+The Kani library needs to be linked as a dev dependency to the crate you're trying to debug.
+This requires adding the following lines to your `Cargo.toml` file,
+which differ depending on what version of the Kani library you would like to use:
+* The latest version:
+```toml
+[dev-dependencies]
```
Comment (celinval) [inline docs/src/debugging-verification-failures.md:21]: Maybe change this to [target.'cfg(Kani)'.dependencies]

So this only gets pulled when user sets `--cfg kani`.

**Review by celinval — APPROVED** [review]
That's great! Minor comments but not blockers.

File: docs/src/concrete-playback.md (hunk 4 above)
Comment (sanjit-bhat) [reply docs/src/concrete-playback.md]: Done

File: docs/src/debugging-verification-failures.md (hunk 11 above)
Comment (sanjit-bhat) [reply docs/src/debugging-verification-failures.md:86]: I'm unclear how to do this? I tried adding this to `Cargo.toml`, but it didn't work:
```toml
[build]
rustflags = ["--cfg=kani"]
```

File: docs/src/debugging-verification-failures.md (hunk 12 above)
Comment (sanjit-bhat) [reply docs/src/debugging-verification-failures.md:94]: Which existing issue (apart from the one I already linked)?

Comment (sanjit-bhat) [reply docs/src/debugging-verification-failures.md:96]: ah, I see. I created a tracking milestone.

File: docs/src/debugging-verification-failures.md (hunk 13 above)
Comment (celinval) [reply docs/src/debugging-verification-failures.md:21]: There are ways around that, but we can tune that later. Thanks

**Pull request merged**